pkg/controller/clone-controller.go

```diff
@@ -45,112 +54,126 @@ const (
-	pvcInformer coreinformers.PersistentVolumeClaimInformer,
-	podInformer coreinformers.PodInformer,
-	image string,
-	pullPolicy string,
-	verbose string,
-	apiServerKey *rsa.PublicKey) *CloneController {
-
-	// Create event broadcaster
-	klog.V(3).Info("Creating event broadcaster")
-	eventBroadcaster := record.NewBroadcaster()
-	eventBroadcaster.StartLogging(klog.V(2).Infof)
-	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: client.CoreV1().Events("")})
-	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: cloneControllerAgentName})
-
-	c := &CloneController{
-		Controller:     *NewController(client, pvcInformer, podInformer, image, pullPolicy, verbose),
-		recorder:       recorder,
+// NewCloneController creates a new instance of the config controller.
+func NewCloneController(mgr manager.Manager, k8sClient kubernetes.Interface, log logr.Logger, image, pullPolicy, verbose string, apiServerKey *rsa.PublicKey) (controller.Controller, error) {
+	reconciler := &CloneReconciler{
+		Client:         mgr.GetClient(),
+		Scheme:         mgr.GetScheme(),
+		Log:            log.WithName("clone-controller"),
 		tokenValidator: newCloneTokenValidator(apiServerKey),
-		cdiClient:      cdiClientSet,
-	}
-	return c
-}
-
-func newCloneTokenValidator(key *rsa.PublicKey) token.Validator {
-	return token.NewValidator(common.CloneTokenIssuer, key, cloneTokenLeeway)
-}
-
-func (cc *CloneController) findCloneSourcePodFromCache(pvc *v1.PersistentVolumeClaim) (*v1.Pod, error) {
-	isCloneRequest, sourceNamespace, _ := ParseCloneRequestAnnotation(pvc)
-	if !isCloneRequest {
-		return nil, nil
+		Image:          image,
+		Verbose:        verbose,
+		PullPolicy:     pullPolicy,
+		recorder:       mgr.GetEventRecorderFor("clone-controller"),
+		K8sClient:      k8sClient,
 	}
-
-	selector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{
-		MatchLabels: map[string]string{
-			CloneUniqueID: string(pvc.GetUID()) + "-source-pod",
-		},
+	cloneController, err := controller.New("clone-controller", mgr, controller.Options{
+		Reconciler: reconciler,
 	})
 	if err != nil {
-		return nil, errors.Wrap(err, "error creating label selector")
+		return nil, err
 	}
-
-	podList, err := cc.podLister.Pods(sourceNamespace).List(selector)
-	if err != nil {
-		return nil, errors.Wrap(err, "error listing pods")
+	if err := addCloneControllerWatches(mgr, cloneController); err != nil {
+		return nil, err
 	}
+	return cloneController, nil
+}
 
-	if len(podList) > 1 {
-		return nil, errors.Errorf("multiple source pods found for clone PVC %s/%s", pvc.Namespace, pvc.Name)
+// addConfigControllerWatches sets up the watches used by the config controller.
+func addCloneControllerWatches(mgr manager.Manager, cloneController controller.Controller) error {
+	// Setup watches
+	if err := cloneController.Watch(&source.Kind{Type: &corev1.PersistentVolumeClaim{}}, &handler.EnqueueRequestForObject{}); err != nil {
```

Comment:
you forgot to watch pods

Reply:
https://github.com/kubevirt/containerized-data-importer/pull/1075/files/d3d13953a54c55c2ebb0bc61e0b7a8a3d4d20dfb#diff-c604e18ed9a6be7674e4881621e1cf6bR109 is where I am watching pods.